### PR TITLE
add fips mode to nginx/passenger

### PR DESCRIPTION
**Pull request by timothy-spencer**
I tested this by applying it by hand.  I will investigate the build process and see if I can get an AMI out of this that I can try in my tspencer env, but for now, I thought I'd put this out for people to comment on.

We appear to be using a FIPS-certified module, but FIPS_mode has not been set.  This turns that on for nginx/passenger

### CODE REVIEW

**Review by pauldoomgov — APPROVED** [review]
Will it FIPS?

LGTM!

**Review by mzia — APPROVED** [review]
LGTM as well

**Comment by timothy-spencer** [comment]
FYI, the nginx/passenger stuff looks good, but the ruby build stuff doesn't seem to be building with the proper FIPS library.  At least it's not passing the tests.  :-(

**Comment by timothy-spencer** [comment]
Hey, I'd love another round of looking at this.  It now passes tests, so I think it's good, but would love to make sure that folks are OK with the openssl changes:
```
    amazon-ebs: Profile: tests from smoke/rails (tests from smoke.rails)
    amazon-ebs: Version: (not specified)
    amazon-ebs: Target:  local://
    amazon-ebs:
    amazon-ebs:   Service ssh
    amazon-ebs:      ✔  is expected to be installed
    amazon-ebs:      ✔  is expected to be enabled
    amazon-ebs:      ✔  is expected to be running
    amazon-ebs:   File /usr/local/bin/id-apt-upgrade
    amazon-ebs:      ✔  is expected to exist
    amazon-ebs:      ✔  is expected to be executable
    amazon-ebs:   File /usr/local/bin/id-chef-client
    amazon-ebs:      ✔  is expected to exist
    amazon-ebs:      ✔  is expected to be executable
    amazon-ebs:   File /usr/local/bin/id-git
    amazon-ebs:      ✔  is expected to exist
    amazon-ebs:      ✔  is expected to be executable
    amazon-ebs:   File /usr/local/bin/git-with-key
    amazon-ebs:      ✔  is expected to exist
    amazon-ebs:      ✔  is expected to be executable
    amazon-ebs:   System Package rbenv
    amazon-ebs:      ✔  is expected to be installed
    amazon-ebs:   File /opt/ruby_build/version
    amazon-ebs:      ✔  is expected to exist
    amazon-ebs:      ✔  content is expected to eq "2.5.8\n"
    amazon-ebs:   Command: `rbenv exec ruby --version`
    amazon-ebs:      ✔  exit_status is expected to eq 0
    amazon-ebs:      ✔  stdout is expected to start with "ruby 2.5.8"
    amazon-ebs:   Command: `rbenv exec ruby -ropenssl -e 'puts OpenSSL::OPENSSL_VERSION'`
    amazon-ebs:      ✔  exit_status is expected to eq 0
    amazon-ebs:      ✔  stdout is expected to start with "OpenSSL 1.0.2u"
    amazon-ebs:   Command: `rbenv exec ruby -ropenssl -e 'OpenSSL.fips_mode = true'`
    amazon-ebs:      ✔  exit_status is expected to eq 0
    amazon-ebs:   Command: `env RBENV_VERSION=2.5 rbenv exec ruby --version`
    amazon-ebs:      ✔  exit_status is expected to eq 0
    amazon-ebs:      ✔  stdout is expected to start with "ruby 2.5"
    amazon-ebs:   Command: `env RBENV_VERSION=2.6 rbenv exec ruby --version`
    amazon-ebs:      ✔  exit_status is expected to eq 0
    amazon-ebs:      ✔  stdout is expected to start with "ruby 2.6"
    amazon-ebs:   Command: `env RBENV_VERSION=2.7 rbenv exec ruby --version`
    amazon-ebs:      ✔  exit_status is expected to eq 0
    amazon-ebs:      ✔  stdout is expected to start with "ruby 2.7"
    amazon-ebs:   File /usr/local/bin/aws
    amazon-ebs:      ✔  is expected to exist
    amazon-ebs:      ✔  is expected to be executable
    amazon-ebs:   Command: `timedatectl status`
    amazon-ebs:      ✔  stdout is expected to include "System clock synchronized: yes"
    amazon-ebs:   File /etc/systemd/timesyncd.conf
    amazon-ebs:      ✔  is expected to exist
    amazon-ebs:      ✔  content is expected to include "NTP=169.254.169.123"
    amazon-ebs:
    amazon-ebs: Test Summary: 30 successful, 0 failures, 0 skipped
```

**Comment by timothy-spencer** [comment]
When launching in my environment, it seems to fail trying to download stuff from s3:
```
              I, [2020-10-19T17:14:33.786223 #10377]  INFO -- deploy/activate: LoginGov::Hostdata::S3: downloading s3://login-gov.app-secrets.XXX-us-west-2/tspencer/idp/v1/application.yml to /srv/idp/releases/chef/config/application_s3_env.yml
              bundler: failed to load command: bin/activate (bin/activate)
              Seahorse::Client::NetworkingError: SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate)
```
Still trying to figure this out.

**Comment by timothy-spencer** [comment]
Fixed the cert verify, but it's now segfaulting randomly while doing the `bundle exec rake db:check_for_pending_migrations` task during activation.  It keeps segfaulting in random places, but pretty much always during something that would be using ssl, like connecting to newrelic or pg.

Initial theories are:
* Before we were using OpenSSL 1.1, which has much better thread safety.  Now we are using the 1.0 FIPS module, which may have less thread safety, as well as require different locking mechanisms.  If this is true, this is going to be basically impossible to make work.
* In straces and the ruby segv handler output, I see mentions of the FIPS 1.0 libcrypto library as well as the system 1.1 library.  Perhaps something is calling into the 1.1 libraries with 1.0 data structures, or the locking systems are incompatible or something.  This seems like it might be more tractable, if we can find out what is loading the system 1.1 lib and tell it to use the 1.0 library instead.

Ugh.  Why can't we use modern crypto implementations?

**Comment by timothy-spencer** [comment]
The only thing I can find with `find /opt/ruby_build/builds/2.6/ -type f | egrep -v '\.rb$' | xargs ldd 2>/dev/null | grep libcrypto.so.1.1` is `/opt/ruby_build/builds/2.6/lib/ruby/gems/2.6.0/gems/passenger-6.0.6/buildout/support-binaries/PassengerAgent`, which is frustrating because it is _supposed_ to be linked properly.  That said, I don't see any mention of that binary being used, so I think it's a red herring.

Still don't know what is loading that lib.  Will continue to fiddle.

**Comment by timothy-spencer** [comment]
removed ruby/gems stuff because it was just broken

File: passenger/recipes/daemon.rb [hunk 1]
```diff
@@ -15,11 +15,34 @@
...
+  patch -p1 < ../fipsmode.patch
+  EOH
+  not_if "test -d #{nginx_path}/src/nginx-#{nginx_version}"
+end
 
 bash "install passenger/nginx" do
   code <<-EOH
   set -eux
-  rbenv exec passenger-install-nginx-module --auto --auto-download --languages ruby --prefix="#{nginx_path}" --extra-configure-flags="#{node[:passenger][:production][:configure_flags]}"
+  rbenv exec passenger-install-nginx-module --auto --nginx-source-dir="#{nginx_path}/src/nginx-#{nginx_version}" --languages ruby --prefix="#{nginx_path}" --extra-configure-flags="#{node[:passenger][:production][:configure_flags]}"
```
Comment (jgrevich) [inline passenger/recipes/daemon.rb:45]: looks fine but I'm curious why we had to change the auto-download flag.

Comment (jgrevich) [reply passenger/recipes/daemon.rb:45]: ahh, probably because we need to first patch the source code before compiling :)

**Review by jgrevich — APPROVED** [review]
Code changes LGTM, I'm a bit concerned about putting this into prod without some solid soak time and perhaps some performance testing.

File: passenger/recipes/daemon.rb (hunk 1 above)
Comment (timothy-spencer) [reply passenger/recipes/daemon.rb:45]: Yup!  You got it!

**Comment by timothy-spencer** [comment]
We have done testing in pt, and this code seems to perform the same as the non-FIPS-mode AMI.

@jgrevich has given me a 👍  on this.

**Comment by jgrevich** [comment]
https://gsa-tts.slack.com/archives/C010L0SE4E8/p1604442863066500

**Pull request merged**